File: hilab-hr/src/lib/scoring.ts
export type Classification = "pass" | "potential" | "fail";

export type ScoreComponents = {
  skills: number;
  experience: number;
  education: number;
  language: number;
};

/** Normalize model-provided scores into the public 0-100 range. */
export function normalizeScore(value: unknown, fallback = 0): number {
  let numeric: number;

  if (typeof value === "number") {
    numeric = value;
  } else if (typeof value === "string") {
    const match = value.match(/\d+(?:\.\d+)?/);
    if (!match) return fallback;
    numeric = Number(match[0]);
    if (numeric > 0 && numeric <= 1) numeric *= 100;
  } else {
    return fallback;
  }

  if (!Number.isFinite(numeric)) return fallback;
  return Math.round(Math.max(0, Math.min(100, numeric)));
}

export interface SkillWeightsInput {
  skills: number;
  experience: number;
  education: number;
  language: number;
}

export function calculateOverallScore(
  scores: ScoreComponents,
  weights?: SkillWeightsInput
): number {
  const sk = normalizeScore(scores.skills);
  const ex = normalizeScore(scores.experience);
  const ed = normalizeScore(scores.education);
  const la = normalizeScore(scores.language);

  if (weights) {
    const total = (weights.skills || 0) + (weights.experience || 0) + (weights.education || 0) + (weights.language || 0);
    if (total > 0) {
      const wSkills = weights.skills / total;
      const wExp = weights.experience / total;
      const wEdu = weights.education / total;
      const wLang = weights.language / total;
      return Math.round(sk * wSkills + ex * wExp + ed * wEdu + la * wLang);
    }
  }

  return Math.round(sk * 0.35 + ex * 0.30 + ed * 0.20 + la * 0.15);
}

export function classifyScore(score: number): Classification {
  const normalized = normalizeScore(score);
  return normalized >= 70 ? "pass" : normalized >= 50 ? "potential" : "fail";
}

File: .agents/skills/hr-cv-screening/scripts/scoring.py
import re


def normalize_score(value, fallback=0) -> int:
    if isinstance(value, bool):
        return fallback
    if isinstance(value, (int, float)):
        number = float(value)
    elif isinstance(value, str):
        match = re.search(r"\d+(?:\.\d+)?", value)
        if not match:
            return fallback
        number = float(match.group(0))
    else:
        return fallback
    if number != number or number in (float("inf"), float("-inf")):
        return fallback
    if 0 < number <= 1:
        number *= 100
    return round(max(0, min(100, number)))


def calculate_overall_score(skills, experience, education, language) -> int:
    return round(
        normalize_score(skills) * 0.35
        + normalize_score(experience) * 0.30
        + normalize_score(education) * 0.20
        + normalize_score(language) * 0.15
    )


def classify_score(score: int) -> str:
    score = normalize_score(score)
    return "pass" if score >= 70 else "potential" if score >= 50 else "fail"
